src/NuGet.Core/NuGet.Versioning/NuGetVersionFactory.cs

```diff
@@ -112,6 +104,168 @@ int IndexOf(string str, char c)
             }
         }
 
+        private static bool TryGetNormalizedVersion(string str, [NotNullWhen(true)] out Version? version)
+        {
+            if (string.IsNullOrWhiteSpace(str))
+            {
+                version = null;
+                return false;
+            }
+
+            int minor = 0;
+            int build = 0;
+            int revision = 0;
+
+            // Check for all the possible parts of the version string. If lastParsedPosition is less than the end of
+            // the string, the input string was invalid (e.g. "1.2.3.4.5").
+            bool success = ParseSection(str, 0, out int lastParsedPosition, out int major) &&
+                ParseSection(str, lastParsedPosition, out lastParsedPosition, out minor) &&
+                ParseSection(str, lastParsedPosition, out lastParsedPosition, out build) &&
+                ParseSection(str, lastParsedPosition, out lastParsedPosition, out revision) &&
+                lastParsedPosition == str.Length;
+
+            if (success)
+            {
+                version = new Version(major, minor, build, revision);
+                return true;
+            }
+            else
+            {
+                version = null;
+                return false;
+            }
+
+            // Returns false if an invalid section was found while processing the string.
+            static bool ParseSection(string str, int start, out int end, out int versionNumber)
+            {
+                // Section is empty.
+                if (start == str.Length)
+                {
+                    end = start;
+                    versionNumber = 0;
+                    return true;
+                }
+
+                // Trim off leading whitespace
+                for (end = start; end < str.Length; ++end)
+                {
+                    char currentChar = str[end];
+                    if (!char.IsWhiteSpace(currentChar))
+                    {
+                        if (IsDigit(currentChar))
```

Comment:
How about using [`Char.IsDigit()`](https://learn.microsoft.com/en-us/dotnet/api/system.char.isdigit?view=net-7.0) API instead of a custom implementation?

Reply:
From: https://github.com/NuGet/NuGet.Client/pull/5202#discussion_r1218465203 (I don't know how well the link will work, since github historically broke deep links to resolved conversations)

> I'd be concerned about unicode characters that return true for `char.IsDigit`, but `(c - '0')` will result in values outside of `0` to `9`.
>
> Having written that, I'm now wondering if that means `Version.Parse` supports other unicode digits, and therefore if we have version strings that are currently parsable that will fail with this change. (though technically I believe those strings would be a violation of the SemVer spec. From memory it's specific about using ASCII '0' to '9'). But, unless the code to convert a single char into a number supports "all" unicode characters, I don't think it's safe to use `char.IsDigit`.

Reply:
IsDigit returns true only for "decimal" digits.

Reply:
IsNumeric is the one that returns true for all unicode numerics.

Reply:
`char.IsDigit('१')` in C# Interactive returns `true`.

Reply:
Indeed, basically this list: https://www.fileformat.info/info/unicode/category/Nd/list.htm. Hmm, I see a few bugs around the tree, assuming this is 0-9. Agreed, probably want to hard code this and add a unit test so it doesn't regress.